Kernel/Devices/TPM/TPM12.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2024, Logkos <65683493+logkos@users.noreply.github.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Types.h>
+#include <Kernel/API/MajorNumberAllocation.h>
+#include <Kernel/Boot/CommandLine.h>
+#include <Kernel/Devices/DeviceManagement.h>
+#include <Kernel/Devices/TPM/Definitions.h>
+#include <Kernel/Devices/TPM/TPM12.h>
+#include <Kernel/Memory/TypedMapping.h>
+
+namespace Kernel {
+
+UNMAP_AFTER_INIT NonnullLockRefPtr<TPMDevice> TPMDevice::create()
+{
+    auto tpm_device_or_error = DeviceManagement::try_create_device<TPMDevice>();
+    VERIFY(!tpm_device_or_error.is_error());
+    auto tpm_device = tpm_device_or_error.release_value();
+    (void)tpm_device->initialize();
+    return *tpm_device;
+}
+
+// FIXME: We assume TPM is always present at the fixed address, which might not be true for all systems.
+//        Ideally, we should discover the TPM using ACPI or platform-specific mechanisms and allow use of other localities
+UNMAP_AFTER_INIT TPMDevice::TPMDevice()
+    : CharacterDevice(MajorAllocation::CharacterDeviceFamily::TPM, 10)
+    , m_registers(Memory::map_typed<TPM12MMIORegistersLocality0 volatile>(PhysicalAddress(0xFED40000), 0x5000, Memory::Region::Access::ReadWrite).release_value_but_fixme_should_propagate_errors())
+{
+}
+
+UNMAP_AFTER_INIT TPMDevice::~TPMDevice() = default;
+
+ErrorOr<void> TPMDevice::initialize()
+{
+    if (!kernel_command_line().is_TPM_enabled())
+        return {};
+
+    TPMPower TPM12Startup;
+    TPM12Startup.header.tag = 0x00C1; // TPM_TAG_RQU_COMMAND
+    TPM12Startup.header.commandSize = sizeof(TPM12Startup);
+    TPM12Startup.header.commandCode = 0x00000099; // TPM_ORD_Startup
+    TPM12Startup.actionType = 0x0001;             // TPM_ST_CLEAR
+
+    TRY(transmit(reinterpret_cast<u8 const*>(&TPM12Startup)));
+    m_initialized = true;
+    dbgln("TPM: Initialized TPM 1.2 device.");
+    return {};
+}
+
+ErrorOr<size_t> TPMDevice::read(OpenFileDescription&, u64, UserOrKernelBuffer& buffer, size_t buffer_size)
+{
+    if (!m_initialized)
+        return Error::from_errno(ENODEV);
+
+    // FIXME: Burst count is not implemented yet
+    Vector<u8> tpm_response;
+    while (m_registers->TPM_STS_0.dataAvail == 0b1) {
+        for (size_t i = 0; i < buffer_size; ++i) {
```

Comment:
I think this might be off, you likely want
```suggestion
        for (size_t i = 0; i < buffer_size && m_registers->TPM_STS_0.dataAvail == 0b1; ++i) 
```
otherwise you might end up endlessly retrieving data

Reply:
(also `FIXME: When being provided a Kernel buffer we could directly write into it`)